### PR TITLE
Add full IIS tests

### PR DESCRIPTION
**Pull request by jkotalik**
A few tests with a full IIS deployer. Making PR to see if jenkins works and blatantly wrong behavior.

### CODE REVIEW

File: src/AspNetCoreModuleV2/RequestHandlerLib/FileOutputManager.cpp [hunk 1]
```diff
@@ -7,7 +7,10 @@
 #include "exceptions.h"
 #include "debugutil.h"
 
-FileOutputManager::FileOutputManager()
+FileOutputManager::FileOutputManager() :
+    m_hLogFileHandle(INVALID_HANDLE_VALUE),
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/RequestHandlerLib/FileOutputManager.cpp]: This is unrelated to this change. Will make it a separate PR.

Comment (pakrym) [reply src/AspNetCoreModuleV2/RequestHandlerLib/FileOutputManager.cpp]: Handle wrapper? :)

Comment (jkotalik) [reply src/AspNetCoreModuleV2/RequestHandlerLib/FileOutputManager.cpp]: Sure.

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                _serverManager.CommitChanges();
+            }
+        }
+
+        private void ConfigureAppPool()
+        {
+            var pool = _serverManager.ApplicationPools.Add(AppPoolName);
+            pool.StartMode = StartMode.AlwaysRunning;
+            pool.ProcessModel.IdentityType = ProcessModelIdentityType.LocalSystem;
+            pool.ProcessModel.LoadUserProfile = true;
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: Why?

File: test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+                var identity = WindowsIdentity.GetCurrent();
+                var principal = new WindowsPrincipal(identity);
+                var isInrole = principal.IsInRole(WindowsBuiltInRole.Administrator);
+
+                var w3wpExists = File.Exists(Path.Combine(Environment.SystemDirectory, "Windows", "System32", "inetsrv", "w3wp.exe"));
+                return w3wpExists && isInrole;
+            }
+        }
+
+        public string SkipReason { get; } = $"Hostable Web Core not availible, {TestServer.HostableWebCoreLocation} not found.";
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs]: Change this to be no w3wp exist or not admin depending on why we are skipping

File: test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                deploymentParameters.ApplicationType = applicationType;
+
+                using (var deployer = new IISDeployer(deploymentParameters, loggerFactory))
+                {
+                    var deploymentRes = await deployer.DeployAsync();
+                    var deploymentResult = new IISDeploymentResult(deploymentRes, logger);
+
+                    if (applicationType == ApplicationType.Standalone)
+                    {
+                        Helpers.ModifyAspNetCoreSectionInWebConfig(deploymentResult, "processPath", "InProcessWebSite.exe");
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: Shouldn't this go into deployer?

File: test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                        Helpers.ModifyAspNetCoreSectionInWebConfig(deploymentResult, "arguments", "");
+                    }
+
+                    // Work around for https://github.com/aspnet/IISIntegration/issues/976
+                    Helpers.ModifyAspNetCoreSectionInWebConfig(deploymentResult, "stdoutLogEnabled", "false");
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: Just set it on web.config

Comment (pakrym) [reply test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: Or I may have done it already

File: test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+                    var response = await RetryHelper.RetryRequest(() =>
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: Retrying client.

File: test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        [SkipIfNotAllowedToRunIIS]
+        public Task HelloWorld_IISExpress_Clr_X64_Portable()
+        {
+            return HelloWorld(RuntimeFlavor.CoreClr, ApplicationType.Portable);
+        }
+
+        private async Task HelloWorld(RuntimeFlavor runtimeFlavor, ApplicationType applicationType)
+        {
+            var testName = $"HelloWorld_{runtimeFlavor}";
+            using (StartLog(out var loggerFactory, testName))
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: Please don't bring the old patterns back.

File: test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs (hunk 5 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: Think you did.

File: test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs (hunk 4 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: Yeah. Lot of this will go away once we move this into hosting.

File: build/dependencies.props [hunk 8]
```diff
@@ -38,6 +38,7 @@
     <MicrosoftNETCoreApp22PackageVersion>2.2.0-preview1-26618-02</MicrosoftNETCoreApp22PackageVersion>
     <MicrosoftNetHttpHeadersPackageVersion>2.2.0-preview1-34530</MicrosoftNetHttpHeadersPackageVersion>
     <MicrosoftNETTestSdkPackageVersion>15.6.1</MicrosoftNETTestSdkPackageVersion>
+    <MicrosoftWebAdministrationPackageVersion>11.1.0</MicrosoftWebAdministrationPackageVersion>
```
Comment (Tratcher) [inline build/dependencies.props]: Is this in our external dependency list in Universe?

File: test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+namespace Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests
+{
+    public class IISTests : LoggedTest
+    {
+        public IISTests(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [Fact]
+        [SkipIfIISExpressSchemaMissingInProcess]
```
Comment (Tratcher) [inline test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: IIS has its own schema file. Do you need a seperate skip attribute?

File: test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs (hunk 3 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs]: copy pasta.

File: test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs (hunk 9 above)
Comment (Tratcher) [reply test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: These Skips can be placed at the class level.

Comment (Tratcher) [reply test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: ConditionalFact

File: test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    public sealed class SkipIfNotAllowedToRunIISAttribute : Attribute, ITestCondition
+    {
+        public bool IsMet {
+            get
+            {
+                // Two conditions need to be met:
+                // IIS needs to be installed
+                var identity = WindowsIdentity.GetCurrent();
+                var principal = new WindowsPrincipal(identity);
+                var isInrole = principal.IsInRole(WindowsBuiltInRole.Administrator);
```
Comment (Tratcher) [inline test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs]: There's a difference between being an administrator and running elevated. Does this detect elevation?

File: test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests
+{
+    [AttributeUsage(AttributeTargets.Assembly | AttributeTargets.Class | AttributeTargets.Method)]
+    public sealed class SkipIfNotAllowedToRunIISAttribute : Attribute, ITestCondition
+    {
+        public bool IsMet {
+            get
+            {
+                // Two conditions need to be met:
```
Comment (Tratcher) [inline test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs]: And running as elevated admin

File: test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs (hunk 7 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: Yeah I'll remove it; I just wanted to get it working first off. Will cleanup now.

File: test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs (hunk 4 above)
Comment (pakrym) [reply test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: Why not now?

Comment (pakrym) [reply test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: We may not move it into hosting.

File: test/IISIntegration.FunctionalTests/Utilities/IISDeployer.cs [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+                var uri = TestIISUriHelper.BuildTestUri(ServerType.IIS, DeploymentParameters.ApplicationBaseUriHint);
+
+                lock (_syncObject)
+                {
+                    // To prevent modifying the IIS setup concurrently.
+                    _application.StartIIS(uri, contentRoot);
+                }
+
+                // Warm up time for IIS setup.
+                await Task.Delay(1000);
```
Comment (Tratcher) [inline test/IISIntegration.FunctionalTests/Utilities/IISDeployer.cs]: 👎 Find a way to check its status.

File: test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+using System;
+using System.IO;
+using System.Security.Principal;
+using Microsoft.AspNetCore.Testing.xunit;
+using Microsoft.Win32;
+
+namespace Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests
+{
+    [AttributeUsage(AttributeTargets.Assembly | AttributeTargets.Class | AttributeTargets.Method)]
+    public sealed class SkipIfNotAllowedToRunIISAttribute : Attribute, ITestCondition
```
Comment (Tratcher) [inline test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs]: SkipIfIISCannotRun?

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+        public string AppPoolName { get; }
+
+        public void StartIIS(Uri uri, string contentRoot)
+        {
+            using (_logger.BeginScope("StartIIS"))
+            {
+                var port = uri.Port;
+                if (port == 0)
+                {
+                    port = (uri.Scheme == "https") ? TestPortHelper.GetNextSSLPort() : TestPortHelper.GetNextPort();
```
Comment (Tratcher) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: This SslPort logic won't work, those ports are specifically bound to IIS Express. You'd have to create an ACL using the dev cert. Throw NotSupported for now?

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+        private void ConfigureAppHostConfig(string dllRoot)
+        {
+            var config = _serverManager.GetApplicationHostConfiguration();
+            var globalModulesSection = config.GetSection("system.webServer/globalModules");
+
+            var ancmFile = GetAncmLocation(dllRoot);
+            var found = false;
+            foreach (var element in globalModulesSection.GetCollection())
+            {
+                if ((string)element["name"] == @"AspNetCoreModuleV2")
```
Comment (Tratcher) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: What about v1? Or is this deployer only for in-proc?

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+            }
+        }
+
+        private void RemoveAppHostConfig()
+        {
+            var config = _serverManager.GetApplicationHostConfiguration();
+            var globalModulesSection = config.GetSection("system.webServer/globalModules");
+            foreach (var element in globalModulesSection.GetCollection())
+            {
+                if ((string)element["name"] == @"AspNetCoreModuleV2")
```
Comment (Tratcher) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: What about the case where AspNetCoreModuleV2 had been added? Should it be removed again?

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+                var addElement = globalModulesSection.GetCollection().CreateElement("add");
+                addElement["name"] = "AspNetCoreModuleV2";
+                addElement["image"] = ancmFile;
+                globalModulesSection.GetCollection().Add(addElement);
+            }
+        }
+
+        private string GetAncmLocation(string dllRoot)
+        {
+            var arch = _deploymentParameters.RuntimeArchitecture == RuntimeArchitecture.x64 ? @"x64\aspnetcorev2.dll" : @"x86\aspnetcorev2.dll";
```
Comment (Tratcher) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: We may have to find a way to prevent any IIS parallel test execution, even across repos.

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs (hunk 16 above)
Comment (pakrym) [reply test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: I think we need to carry our own apphostconfig backup old one and use ours for tests to avoid machine state affecting tests.

File: test/IISIntegration.FunctionalTests/Utilities/IISDeployer.cs [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+                    contentRoot = DeploymentParameters.PublishedApplicationRootPath;
+                    //AddAccessPermissions(contentRoot, _application.AppPoolName);
+                }
+
+                var uri = TestIISUriHelper.BuildTestUri(ServerType.IIS, DeploymentParameters.ApplicationBaseUriHint);
+
+                lock (_syncObject)
+                {
+                    // To prevent modifying the IIS setup concurrently.
+                    _application.StartIIS(uri, contentRoot);
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Utilities/IISDeployer.cs]: Did you check if running other tests concurrently would kill already running apps?

File: test/IISIntegration.FunctionalTests/Utilities/IISDeployer.cs (hunk 12 above)
Comment (pakrym) [reply test/IISIntegration.FunctionalTests/Utilities/IISDeployer.cs]: I thought management API had a way to check status.

File: test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs (hunk 4 above)
Comment (Tratcher) [reply test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: Let's try keeping in here and see if it's easier to maintain. However you'd need to add a separate package if you want it to be consumed in other repos. That can wait though.

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs (hunk 16 above)
Comment (Tratcher) [reply test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: That's dramatic. Even more reason to work out a way to prevent parallel execution across repos.

File: test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs (hunk 4 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Inprocess/IISTests.cs]: Alright, I'll err on the side of putting everything in here then.

File: build/dependencies.props (hunk 8 above)
Comment (jkotalik) [reply build/dependencies.props]: Believe we will need to add it.

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+        }
+
+        private void RemoveAppPool()
+        {
+            var appPool = _serverManager.ApplicationPools.FirstOrDefault(pool => pool.Name == AppPoolName);
+            if (appPool != null)
+            {
+                appPool.Stop();
+                _serverManager.ApplicationPools.Remove(appPool);
+                _serverManager.CommitChanges();
```
Comment (mikeharder) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: When adding the site and app pool, the code calls `_serverManager.CommitChanges()` once at the end.  But when removing the site and app pool, the code calls `_serverManager.CommitChanges()` after each operation.  These should be consistent, and I would change the removal code to call once at the end if it's sufficient.  Minimize calls to `CommitChanges()`.

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+            }
+        }
+
+        private void ConfigureAppPool()
+        {
+            var pool = _serverManager.ApplicationPools.Add(AppPoolName);
+            pool.StartMode = StartMode.AlwaysRunning;
+            pool.ProcessModel.IdentityType = ProcessModelIdentityType.LocalSystem;
+            pool.ProcessModel.LoadUserProfile = true;
+        }
```
Comment (mikeharder) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: [Docs](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/iis/?view=aspnetcore-2.1&tabs=aspnetcore2x#create-the-iis-site) recommend setting the CLR version to "No Managed Code".  I think we should do the same, by setting `pool.ManagedRuntimeVersion = string.Empty;`

**Comment by jkotalik** [comment]
Still working out parallel test execution problems a few newer comments. If we want to keep the IISDeployer in this repo, we should move the IISExpress deployer here too. There are too many shared components between them.

File: test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs (hunk 10 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs]: It doesn't. Not sure if we care though?

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs (hunk 17 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: I'd handle this in another PR as this is the only repo doing it.

**Comment by jkotalik** [comment]
It seems IIS in installed on appveyor. Seems to be timing out though; will look into why.

File: test/IISIntegration.FunctionalTests/Utilities/IISDeployer.cs (hunk 18 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Utilities/IISDeployer.cs]: Tests are already running sequentially

**Comment by jkotalik** [comment]
I think I hit most of the feedback. 🆙 📅

File: test/IISIntegration.FunctionalTests/Utilities/IISAncmValidator.cs [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+                SkipReason = "Could not read ANCM schema configuration";
+                return;
+            }
+
+            IsMet = ancmConfig
+                .Root
+                .Descendants("attribute")
+                .Any(n => "hostingModel".Equals(n.Attribute("name")?.Value, StringComparison.Ordinal));
+
+            SkipReason = IsMet ? null : "IIS schema be upgraded to support ANCM.";
```
Comment (mikeharder) [inline test/IISIntegration.FunctionalTests/Utilities/IISAncmValidator.cs]: Typo in string

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+    {
+        private const int IISRetryLimit = 500;
+        private const int IISWaitMillisecondsTimeout = 10;
+        private readonly ServerManager _serverManager = new ServerManager();
+        private readonly DeploymentParameters _deploymentParameters;
+        private readonly ILogger _logger;
+        private readonly string _ancmVersion;
+        private readonly object _syncLock = new object();
+        private readonly string _apphostConfigPath = Path.Combine(Environment.SystemDirectory, "inetsrv", "config", "applicationhost.config");
+        private readonly string _apphostConfigBackupPath = Path.Combine(Environment.SystemDirectory, "inetsrv", "config", "applicationhost.config.backup");
```
Comment (mikeharder) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: Change to `private static readonly`

File: test/IISIntegration.FunctionalTests/Utilities/SkipIfIISCannotRunAttribute.cs [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+using Microsoft.AspNetCore.Server.IntegrationTesting;
+using Microsoft.AspNetCore.Testing.xunit;
+using Microsoft.Win32;
+
+namespace Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests
+{
+    [AttributeUsage(AttributeTargets.Assembly | AttributeTargets.Class | AttributeTargets.Method)]
+    public sealed class SkipIfIISCannotRunAttribute : Attribute, ITestCondition
+    {
+
```
Comment (mikeharder) [inline test/IISIntegration.FunctionalTests/Utilities/SkipIfIISCannotRunAttribute.cs]: Remove extra whitespace

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 24]
```diff
@@ -0,0 +1,249 @@
...
+
+namespace Microsoft.AspNetCore.Server.IntegrationTesting
+{
+    /// <summary>
+    /// Represents the IIS website registered in the global applicationHost.config
+    /// </summary>
+    internal class IISApplication
+    {
+        private const int IISRetryLimit = 500;
+        private const int IISWaitMillisecondsTimeout = 10;
```
Comment (mikeharder) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: I think it's a bit cleaner to instead use two `TimeSpan`:

```
private static readonly TimeSpan _timeout = TimeSpan.FromSeconds(5);
private static readonly TimeSpan _retryDelay = TimeSpan.FromMilliseconds(10);
...
var sw = Stopwatch.StartNew();
while (sw.Elapsed < _timeout) {
    // Attempt to check status, return if done
    await Task.Delay(_retryDelay);
}
```

**Comment by pakrym** [comment]
Make appveyor pass.

File: test/IISIntegration.FunctionalTests/Utilities/IISAncmValidator.cs [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+            {
+                SkipReason += "The machine does not have IIS installed.";
+                return;
+            }
+
+            var ancmConfigPath = Path.Combine(Environment.SystemDirectory, "inetsrv", "config", "schema", "aspnetcore_schema_v2.xml");
+
+            if (!File.Exists(ancmConfigPath))
+            {
+                SkipReason = "IIS is not installed.";
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Utilities/IISAncmValidator.cs]: Schema

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 26]
```diff
@@ -0,0 +1,249 @@
...
+        private readonly ServerManager _serverManager = new ServerManager();
+        private readonly DeploymentParameters _deploymentParameters;
+        private readonly ILogger _logger;
+        private readonly string _ancmVersion;
+        private readonly object _syncLock = new object();
+        private readonly string _apphostConfigPath = Path.Combine(Environment.SystemDirectory, "inetsrv", "config", "applicationhost.config");
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: Are we planning on doing x64/x86 testing?

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 27]
```diff
@@ -0,0 +1,249 @@
...
+            RestoreAppHostConfig();
+
+            _serverManager.CommitChanges();
+
+            await WaitUntilSiteStopped();
+        }
+
+        private async Task WaitUntilSiteStopped()
+        {
+            var i = 0;
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: move this into a loop and use IISRetryLimit for logs

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 28]
```diff
@@ -0,0 +1,249 @@
...
+
+        private void AddTemporaryAppHostConfig()
+        {
+            File.Copy(_apphostConfigPath, _apphostConfigBackupPath);
+            _logger.LogInformation($"Backed up {_apphostConfigPath} to {_apphostConfigBackupPath}");
+        }
+
+        private void RestoreAppHostConfig()
+        {
+            File.Delete(_apphostConfigPath);
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: Historically delete always needs a retry.

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs (hunk 22 above)
Comment (pakrym) [reply test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: Add a random part to it, if tests crashes subsequent backup would override original backup and file will be lost.

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 29]
```diff
@@ -0,0 +1,249 @@
...
+            {
+                _logger.LogInformation($"Could not find {_ancmVersion} section in global modules; creating section.");
+                var addElement = globalModulesSection.GetCollection().CreateElement("add");
+                addElement["name"] = _ancmVersion;
+                addElement["image"] = ancmFile;
+                globalModulesSection.GetCollection().Add(addElement);
+            }
+            else
+            {
+                _logger.LogInformation($"Replacing {_ancmVersion} section in global modules");
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: with what?

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs (hunk 22 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: I think the site name needs to have more randomness. It is based off of current time up to the nearest second.

File: test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs (hunk 10 above)
Comment (Tratcher) [reply test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs]: Won't the test fail without elevation?

**Comment by Tratcher** [comment]
What's shared with the express deployer?

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs (hunk 26 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: Yes, but the config is shared across 32 and 64 bit for IIS.

**Comment by jkotalik** [comment]
> What's shared with the express deployer?

Modifying apphost.config stuff, enabling inplace test running, ANCM finding, etc.

**Comment by Tratcher** [comment]
Not enough to block a first pass. We can work out homes once we prove this one is working.

**Comment by jkotalik** [comment]
Investigating appveyor failures. Also trying on a feature branch.

**Comment by natemcmaster** [comment]
Can you hold off for an hour or so until I can finish 2.2 branching changes? Also, you may want to retarget this to release/2.2

File: test/IISIntegration.FunctionalTests/Utilities/IISAncmValidator.cs [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                SkipReason = "IIS tests can only be run on Windows";
+                return;
+            }
+
+            var identity = WindowsIdentity.GetCurrent();
+            var principal = new WindowsPrincipal(identity);
+            if (!principal.IsInRole(WindowsBuiltInRole.Administrator))
```
Comment (Tratcher) [inline test/IISIntegration.FunctionalTests/Utilities/IISAncmValidator.cs]: I didn't see an answer to the elevation question. Do the tests pass without elevation?

File: test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs (hunk 10 above)
Comment (Tratcher) [reply test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs]: You cannot enable these tests if they're going to fail when running non-elevated. Full Stop.

Is that why the AppVeyor tests are failing?

Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Utilities/SkipIfNotAllowedToRunIISAttribute.cs]: Resolved offline. WindowsBuiltInRole.Administrator returns true only if your current role is elevated. The skip condition works for skipping builds when running non-admin.

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 31]
```diff
@@ -199,20 +199,22 @@ private ApplicationPool ConfigureAppPool(string contentRoot)
...
-            var addElement1 = envCollection.CreateElement("add");
-            addElement1["name"] = @"ASPNETCORE_MODULE_DEBUG";
-            addElement1["value"] = @"4";
-            envCollection.Add(addElement1);
-            var addElement2 = envCollection.CreateElement("add");
-            addElement2["name"] = @"ASPNETCORE_MODULE_DEBUG_FILE";
-            addElement2["value"] = Path.Combine(contentRoot, $"{WebSiteName}.txt");
-            envCollection.Add(addElement2);
+
+            AddEnvironmentVariableToAppPool(envCollection, @"ASPNETCORE_MODULE_DEBUG", @"4");
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: What about other deployment options env variables?

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs [hunk 32]
```diff
@@ -0,0 +1,303 @@
...
+            {
+                _logger.LogInformation($"Replacing {_ancmVersion} section in global modules with {ancmFile}");
+                globalConfigElement["image"] = ancmFile;
+            }
+        }
+
+        private void SetModulesSection(Configuration config)
+        {
+            var modulesSection = config.GetSection("system.webServer/modules");
+            var moduleConfigElement = modulesSection.GetCollection().Where(element => (string)element["name"] == _ancmVersion).FirstOrDefault();
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: What if you found it but it was wrong?

File: test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs (hunk 17 above)
Comment (jkotalik) [reply test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs]: Across repos is a non-goal for right now.
